mrmustard/physics/ansatze.py

```diff
@@ -241,6 +259,76 @@ def _order_batch(self):
         self.vec = math.gather(self.vec, sorted_indices, axis=0)
         self.array = math.gather(self.array, sorted_indices, axis=0)
 
+    def decompose_ansatz(self) -> DiffOpPolyExpAnsatz:
+        r"""
+        This method decomposes a DiffOpPolyExpAnsatz. Given an ansatz of dimensions:
+        A=(batch,m+n,m+n), b=(batch,m+n), c = (batch,k_1,k_2,...,k_n),
+        it can be rewritten as an ansatz of dimensions
+        A=(batch,2m,2m), b=(batch,2m), c = (batch,l_1,l_2,...,l_m), with l_i = sum_j k_j
+        This decomposition is typically favourable if n>m, and will only run if that is the case.
+        """
+        dim_beta, shape_beta = self.polynomial_degrees
+        dim_alpha = self.mat.shape[-1] - dim_beta
+        batch_size = self.batch_size
+        if dim_beta > dim_alpha:
+            A_bar = np.array(
+                [
+                    np.block(
+                        [
+                            [
+                                np.zeros((dim_alpha, dim_alpha)),
+                                self.mat[i, dim_alpha:, :dim_alpha].T,
+                            ],
+                            [
+                                self.mat[i, dim_alpha:, :dim_alpha],
+                                self.mat[i, dim_alpha:, dim_alpha:],
+                            ],
+                        ]
+                    )
+                    for i in range(batch_size)
+                ]
+            )
+
+            b_bar = np.array(
+                [
+                    np.concatenate((np.zeros(dim_alpha), self.vec[i, dim_alpha:]))
+                    for i in range(batch_size)
+                ]
+            )
+
+            poly_bar = math.hermite_renormalized_batch(
```

Comment:
Given that this is the only place in MM where this function is used, it may be worth modifying its definition to work with the batch at the beginning so that we avoid the moveaxis calls back and forth. The function is called `vanilla_batch`, it's super easy to modify. Also, we should use `math` rather than `np` whenever the function should be differentiable with respect to at least one argument (e.g. `np.sum`, `np.block`, etc... break the chain rule because they should be differentiable. Use `math.sum` `math.block` etc. (We don't support every possible numpy function and sometimes the signatures differ)

Reply:
Do you think we can rewrite it even further so that it can work with two different batch dimensions, one for the argument and one for the Abc? Then we can avoid the loop around it.